[bokun] iP

bokung opened the pull request:
### UwUke
> "OwO" - UwU [sauce](https://youtu.be/dQw4w9WgXcQ)

UwU
- *OwO*
- **OwO OwO**
- OwO OwO OwO

(>w<)
1. ^w^ :robot:
2. ^w^ :robot:
3. ^w^ :robot:

YwY
- [x] ~~UwU~~ VwV
- [ ] VwV
- [ ] VwV

Here is some code to perform a `binarySearch`, but there's a bug hidden somewhere! Can you spot what's wrong?
``` java
public class BinarySearch {
    public static int binarySearch(int[] elements, int target) {
        int begin = 0;
        int end = elements.length - 1;
        while (begin < end) {
            int mid = begin + (end-begin)/2;
            if (target <= elements[mid]) {
                end = mid;
            } else {
                begin = mid;
            }
        }
        return elements[begin] == target ? begin : -1;
    }
}
```

Commit:
Add Gradle support

bokung committed:
ignore .class files

bokung committed:
Level 1 done

bokung committed:
close scanner

bokung committed:
add decorations

bokung committed:
group stuff

bokung committed:
Level 2

bokung committed:
Create Task.java

bokung committed:
add regex to detect mark and unmark

bokung committed:
Update Duke.java

bokung committed:
Level 3

bokung committed:
add task children classes

bokung committed:
refactor

bokung committed:
Create Command.java

bokung committed:
command checker

bokung committed:
refactor again

bokung committed:
add input parser

bokung committed:
Level 4

bokung committed:
move command to right place

bokung committed:
refactor, remove original add

bokung committed:
add tests

bokung committed:
handle errors

bokung committed:
refactor to use arraylist

bokung committed:
add delete

bokung committed:
forgot to print delete message

bokung committed:
enhanced duke's name

bokung committed:
Update .gitignore to ignore data folder

bokung committed:
Move DukeException to output package

bokung committed:
Add ability to save tasks between runs

bokung committed:
Treat deadlines as dates

bokung committed:
Merge branch 'branch-Level-8'

bokung committed:
Merge branch 'add-gradle-support'

bokung committed:
Update build.gradle

bokung committed:
Create TaskList.java

bokung committed:
Clean up main

bokung committed:
Move everything to one package

bokung committed:
Add basic tests

bokung committed:
Add some documentation

bokung committed:
Capitalize constants

bokung committed:
Add find

bokung committed:
Merge branch 'branch-A-CodingStandard'

bokung committed:
Merge branch 'branch-Level-9'

bokung committed:
Fix long line

bokung committed:
Fix indentation

bokung committed:
Clean up main a bit

russelltankaimin commented on src/main/java/task/Task.java:
```diff
@@ -8,27 +8,24 @@ public Task(String taskName) {
     }
 
     public String markDone() {
         this.markedDone = true;
```
I think a better naming for `markedDone` boolean variable would be `isDone`. According to the textbook, some of the recommended prefixed are `is`, `has` etc. For your consideration 😊

bokung replied:
Good catch, will update in next version!

russelltankaimin commented on src/main/java/Duke.java:
```diff
@@ -28,9 +28,7 @@ private static void bye() {
     private static void displayTasks(Task[] tasks, int numTasks) {
         StringBuilder sb = new StringBuilder();
```
I might be a bit nitpicky, but perhaps a better naming for this String Building could just simply be `stringBuilder` rather than `sb`. For your consideration 😊

bokung replied:
Will consider! Though for this one the scope is rather limited and this stringbuilder is not really being used outside of this function.

russelltankaimin commented on src/main/java/task/Task.java:
```diff
@@ -8,27 +8,24 @@ public Task(String taskName) {
+        return String.format("%s %s", this.getStatusIcon(), this.taskName);
     }
 
     public String markDone() {
```
Yes, one good habit in your coding style is that you name methods starting with verbs! Great job! 👍

russelltankaimin reviewed (commented):
Great job overall with your overall coding style. I do not see any major issues. You followed almost everything from the coding style guide.

russelltankaimin commented on src/main/java/Duke.java:
```diff
@@ -18,7 +18,7 @@ public class Duke {
 
     private final static int CAPACITY = 100;
```
I think the order should have been `private static final` instead.

bokung replied:
Will take note.

russelltankaimin commented on src/main/java/Duke.java:
```diff
@@ -94,14 +96,24 @@ public static void main(String[] args) {
                         break;
                     case MARK:
```
Do make sure, your `switch` and `case` keywords lie under the same indentation line 👌

bokung replied:
Good catch. Will update in future version.

panavdua commented on src/main/java/uwuke/input/Advisor.java:
```diff
@@ -0,0 +1,26 @@
+     * @param input faulty input string given by user
+     * @return advise based on faulty input
+     */
+    public static String advise(String input) {
```
Perhaps here you could make the name of the function more indicative of what it is doing, for example - `adviseUsers`

bokung replied:
Will do that in the next version! It does indeed appear that this entire class is a bit ambiguous and can be named to reflect its purpose better.

panavdua commented on src/main/java/uwuke/input/Parser.java:
```diff
@@ -0,0 +1,149 @@
+     * @return string that can be used directly to create a new deadline object
+     * @throws DukeException if input cannot be parsed
+     */
+    public static String parseTodo(String input) throws DukeException {
```
Very well-written function!! But wanted to ask if you would prefer to keep `Todo` or change it to `ToDo`?

bokung replied:
Yes it should be `ToDo`. Good catch!

panavdua reviewed (commented):
You really have adhered closely to the coding standard, could not find any major error!!

bokung committed:
Add basic window

bokung committed:
Add GUI

bokung committed:
Delete Main.java

bokung committed:
Make code more readable

bokung committed:
Add flipped image

bokung committed:
Shift image handling to DialogBox

bokung committed:
Minor fixes

bokung committed:
Fix indentation

bokung committed:
Fix failing initialisation

bokung committed:
Clean up

bokung committed:
Rename boolean

bokung committed:
Fix documentation and method names

bokung committed:
Minor name change

bokung committed:
Remove unnecessary import

ParserTest originally included an import for Parser.

This is actually not needed, since the SUT is imported automatically

bokung committed:
Add assertions for GUI components

Include more assert statements

When handling User Inputs, relevant GUI components should be initialised or unexpected runtime errors can occur.

Adding assert statements before calling GUI component methods can catch null pointers before they occur, and give useful information to fix the issue swiftly.

bokung committed:
Add assertions for image resources

DialogBox currently is exceedingly reliant on the resource loader class providing the correct image.

Class path issues can potentially cause the resource loader to fail in certain environments, and the DialogBox class would have no access to images that are required to construct dialog boxes.

Adding assertions during this initialisation stage can help catch potential bugs when the image is not loaded.

bokung committed:
Add assertions for tasks

There are checks to ensure the capacity of task lists are not violated.

However there is currently no easy way to know when a task list is overloaded past its capacity.

Adding assertions will help with enforcing the invariant that the number of elements in the task list should not exceed the capacity.

bokung committed:
Refactor start method

Main start method for JavaFX currently has two layers of abstraction.

This does not provide a good high level overview of what the method is supposed to do, considering how high level the start method is.

Extracting and grouping parts of the start function into new methods helps to reduce the layers of abstraction to one, and makes the method much easier to read.

bokung committed:
Add minor fix

bokung committed:
Merge pull request #2 from bokung/branch-A-Assertions

Branch a assertions

bokung committed:
Merge pull request #3 from bokung/branch-A-CodeQuality

Branch a code quality

bokung committed:
Major reorganisation of packages

Some classes are not grouped by concrete functionality.

Packages should group classes that interact heavily with each other together.

Moving classes according to the MVC architecture makes the overall structure much more prominent and easy to understand

bokung committed:
Fix formatting

bokung committed:
Add ability to check if deadline is due soon

bokung committed:
Add reminder for deadlines

bokung committed:
Set up CI

bokung committed:
Refactor to use FXML for GUI

bokung committed:
Minor refactor

bokung committed:
Change colour scheme

bokung committed:
Added a weird looking button

bokung committed:
Change button color

bokung committed:
White text

bokung committed:
Change dialog box colors

bokung committed:
dont flip

bokung committed:
Disable resizing

bokung committed:
Create Ui.png

bokung committed:
Fix reminders not loading properly

bokung committed:
Merge branch 'branch-A-BetterGui'

bokung committed:
Use new logo

bokung committed:
Give user inputs different color

bokung committed:
Add new image

bokung committed:
Remove ubuntu

bokung committed:
Revert "Remove ubuntu"

This reverts commit 8b10e970a0c71471f13c2d836c58abe44fe5a543.

bokung committed:
Fix button

bokung committed:
Remove unused imports

bokung committed:
Reduce icon size

bokung committed:
Add User Guide

bokung committed:
Fix truncated content

bokung committed:
Update build.gradle

bokung committed:
Final UI fixes
